sdk/core/core/src/az_span.c

```diff
@@ -646,10 +646,24 @@ AZ_NODISCARD az_result _az_span_url_encode(az_span destination, az_span source,
   int32_t result_size = 0;
   for (int32_t i = 0; i < source_size; ++i)
   {
-    result_size += _az_span_url_should_encode(az_span_ptr(source)[i]) ? 3 : 1;
+    if (_az_span_url_should_encode(az_span_ptr(source)[i]))
+    {
+      if (result_size <= INT32_MAX - 3)
```

Comment:
I have a feeling we can re-write this with less overhead.

Consider also adding a precondition that source cannot be `> INT_MAX/3`. I don't think we need to support anything larger.

Maybe you can figure out ways to reduce the number of checks happening within a loop. If I have some suggestion, I will let you know.

Reply:
Again, if you have more concrete suggestions, please send them to me. I had a couple of ideas, but many of them ending up relying on adding up two integers, which we especially want to avoid in this case, or this entire check can't be trusted.
I would try not to disallow INT32_MAX/3 if possible, that's about 715MB, large but still relatively reasonable number.
I also do not keep the goal to optimize "fail fast" scenario, as this might never happen.
Another thought, if you run this multithreaded, and change contents of the source buffer (not the size), then the size of the encoded result we calculated and checked won't be sufficient, we still may write outside of the destination span. To guard against that, we have to always require destination being 3x larger than the source.
I'm gonna think about that over the weekend.

Reply:
This all reminds me... At a time, we used to have az_assumptions.c - a file that we compile and that produces no output, but it was checking all our assumptions at a compile time, so everywhere in the code we would know whether it is safe rely on some behavior, or not. And if that file won't compile, our SDK won't work correctly on that platform, and needs to be updated: https://github.com/Azure/azure-sdk-for-c/blob/669fe8ca88b3defb1d8ad4a9c8b69ced8b6f3267/sdk/core/core/src/az_assumptions.c

We could've write static assert of (INT32_MAX + 1 == INT32_MIN in that file, or maybe right above the function, so that if that breaks at compile time, you know that you need to update the block that checks destination span size, or remove it entirely at your own risk) - with that assert we could write much simper check of size being negative, and it would probably work on 100% of platforms our SDK ever gets compiled on.

Reply:
> with that assert we could write much simper check of size being negative, and it would probably work on 100% of platforms our SDK ever gets compiled on.

That wouldn't completely work either. Since we are multiplying by 3, it could overflow back to a positive number and still result in data loss.

Reply:
Check out the new commit, I think it is less overhead, I like it.

Reply:
I wrote yet another version, with size calculation extracted to another (inline) function, which has the loop code duplicated a bit, but it should fail fast / work bit faster in cases when there's we know upfront there's enough space.

Reply:
If we remove the guarantee which we never explicitly gave, that we won't modify destination if result can't fit, we can make everything faster with just one pass when reading the source. Which makes a lot of sense to do IF/when this code needs to process tons of data. Plus, it will behave a bit better if the source buffer gets modified by another thread. The nature of algorithm is that to get that final size is to scan all the input once. It seems like a reasonable tradeoff, I don't see much code that would be unavoidably inconvenienced by this "relaxed" contract.

Reply:
What's good is that for the cases when you provide destination buffer which is 3x in size, you pay almost no correctness check penalty. This technically can be achieved today, even with the upfront check size check, but in the case when destination is less than 3x of source, we scan the source twice.

Reply:
Yes, I went all in with that idea, it is yet another version now. If needed, I can roll back a few commits, but I think it is reasonable to not scan twice on IOT devices for the small tradeoff which can be avoided.

Reply:
> I would try not to disallow INT32_MAX/3 if possible, that's about 715MB, large but still relatively reasonable number.

As an aside, I really don't think 715MB is too small a limitation for a URL. In this context, I think we can totally limit us to a much smaller number, if we really needed to.

Reply:
> If we remove the guarantee which we never explicitly gave, that we won't modify destination if result can't fit, we can make everything faster with just one pass when reading the source.

I think it is fine for us to remove that guarantee.

Reply:
Completed.